pallets/external-validator-slashes/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+        NotSortedAndUnique,
+        ProvidedFutureEra,
+        ProvidedNonSlashableEra,
+        ActiveEraNotSet,
+        DeferPeriodIsOver,
+        ErrorComputingSlash,
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    /// All slashing events on validators, mapped by era to the highest slash proportion
+    /// and slash value of the era.
+    #[pallet::storage]
+    pub(crate) type ValidatorSlashInEra<T: Config> =
+        StorageDoubleMap<_, Twox64Concat, EraIndex, Twox64Concat, T::AccountId, Perbill>;
+
+    /// A mapping from still-bonded eras to the first session index of that era.
+    ///
+    /// Must contains information for eras for the range:
+    /// `[active_era - bounding_duration; active_era]`
+    #[pallet::storage]
+    #[pallet::unbounded]
+    pub type BondedEras<T: Config> = StorageValue<_, Vec<(EraIndex, SessionIndex)>, ValueQuery>;
+
+    /// A counter on the number of slashes we have performed
+    #[pallet::storage]
+    #[pallet::getter(fn next_slash_id)]
+    pub type NextSlashId<T: Config> = StorageValue<_, T::SlashId, ValueQuery>;
+
+    /// All unapplied slashes that are queued for later.
+    #[pallet::storage]
+    #[pallet::unbounded]
+    #[pallet::getter(fn slashes)]
+    pub type Slashes<T: Config> =
+        StorageMap<_, Twox64Concat, EraIndex, Vec<Slash<T::AccountId, T::SlashId>>, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Cancel a slash that was deferred for a later era
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::cancel_deferred_slash(slash_indices.len() as u32))]
+        pub fn cancel_deferred_slash(
+            origin: OriginFor<T>,
+            era: EraIndex,
+            slash_indices: Vec<u32>,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            let active_era = T::EraIndexProvider::active_era().index;
+
+            // We need to be in the defer period
+            ensure!(
+                era <= active_era
+                    .saturating_add(T::SlashDeferDuration::get().saturating_add(One::one()))
+                    && era > active_era,
+                Error::<T>::DeferPeriodIsOver
+            );
+
+            ensure!(!slash_indices.is_empty(), Error::<T>::EmptyTargets);
+            ensure!(
+                is_sorted_and_unique(&slash_indices),
+                Error::<T>::NotSortedAndUnique
+            );
+            // fetch slashes for the era in which we want to defer
+            let mut era_slashes = Slashes::<T>::get(&era);
+
+            let last_item = slash_indices[slash_indices.len() - 1];
+            ensure!(
+                (last_item as usize) < era_slashes.len(),
+                Error::<T>::InvalidSlashIndex
+            );
+
+            for (removed, index) in slash_indices.into_iter().enumerate() {
+                let index = (index as usize) - removed;
+                era_slashes.remove(index);
+            }
```

Comment:
This is a weird way to remove multiple elements from a vec. I know it's copied from polkadot but just do this instead:

```suggestion
            // Remove elements starting from the highest index to avoid shifting issues.
            for index in slash_indices.into_iter().rev() {
                era_slashes.remove(index);
            }
```

Performance could be improved by using Vec::retain, but that would be too complex.

Reply:
fixed